hw/ip/otbn/data/insns.yml

```diff
@@ -29,6 +29,353 @@ insn-groups:
+
+  # RISC-V "I-type" encoding (reg <- fun(imm, reg))
+  I:
+    parents:
+      - rv
+      - funct3
+    fields:
+      imm: 31-20
+      rs1: 19-15
+      rd: 11-7
+
+  # RISC-V "S-type" encoding (_ <- fun(reg, imm))
+  S:
+    parents:
+      - rv
+      - funct3
+    fields:
+      imm: 31-25,11-7
+      rs2: 24-20
+      rs1: 19-15
+
+  # RISC-V "B-type" encoding (like S, but different immediate layout; used for
+  # branches)
+  B:
+    parents:
+      - rv
+      - funct3
+    fields:
+      imm:
+        bits: 31,7,30-25,11-8
+        shift: 1
+      rs2: 24-20
+      rs1: 19-15
+
+  # RISC-V "U-type" encoding (reg <- fun(imm))
+  U:
+    parents:
+      - rv
+    fields:
+      imm:
+        bits: 31-12
+        shift: 12
+      rd: 11-7
+
+  # RISC-V "J-type" encoding (like U, but different immediate layout; used for
+  # jumps)
+  J:
+    parents:
+      - rv
+    fields:
+      imm:
+        bits: 31,19-12,20,30-21
+        shift: 1
+      rd: 11-7
+
+  # A partial scheme for custom instructions with opcode b00010
+  custom0:
+    parents:
+      - rv(opcode=b00010)
+
+  # A partial scheme for custom instructions with opcode b01010
+  custom1:
+    parents:
+      - rv(opcode=b01010)
+
+  # A partial scheme for custom instructions with opcode b01110
+  custom2:
+    parents:
+      - rv(opcode=b01110)
+
+  # A partial scheme for custom instructions with opcode b11110
+  custom3:
+    parents:
+      - rv(opcode=b11110)
+
+  # A partial scheme for bignum instructions that produce a dest WDR.
+  wrd:
+    fields:
+      wrd: 11-7
+
+  # A partial scheme for bignum instructions that take two source WDRs and
+  # produce a dest WDR.
+  wdr3:
+    parents:
+      - wrd
+    fields:
+      wrs2: 24-20
+      wrs1: 19-15
+
+  # A partial scheme that defines the 'fg' field (for <flag_group> operands)
+  fg:
+    fields:
+      fg: 31
+
+  # A partial scheme that defines the shift fields (type and bytes)
+  shift:
+    fields:
+      shift_type: 30
+      shift_bytes: 29-25
+
+  # A partial scheme that defines a function field at bit 31 for OTBN logical
+  # operations
+  funct31:
+    fields:
+      funct31: 31
+
+  # A partial scheme for specialized 2 bit function field, we need a reduced
+  # size in the lower two bits of funct3 as RSHI spills over 1 bit from its
+  # immediate
+  funct2:
+    fields:
+      funct2: 13-12
+
+  # A specialised encoding for the loop instruction (only one source, no
+  # destination)
+  loop:
+    parents:
+      - custom3
+      - funct2(funct2=b00)
+    fields:
+      bodysize: 31-20
+      grs: 19-15
+      fixed:
+        bits: 14,11-7
+        value: bxxxxxx
+
+  # A specialised encoding for the loopi instruction (which, unusually, has 2
+  # immediates)
+  loopi:
+    parents:
+      - custom3
+      - funct2(funct2=b01)
+    fields:
+      bodysize: 31-20
+      iterations: 19-15,11-7
+      fixed:
+        bits: 14
+        value: bx
+
+  # Used for bignum logical operations (and, or, xor).
```

Comment:
Let's call those "BN": BN.AND, BN.OR, BN.XOR. And, in line with the RISC-V Spec, instructions are all-caps.

Reply:
Done.

Reply:
Oh, I imagined that this would be case insensitive. Do you think we really have to shout?

Reply:
I don't have a strong style preference either way, but keeping things consistent with the RISC-V Spec seems helpful to me. Up to you, though.

Reply:
Let's be case insensitive for now (to preserve my delicate eyes :-D). We can always do a mega find-and-replace in the future if it gets confusing.